src/main/java/uk/gov/justice/hmpps/datacompliance/services/duplicate/detection/image/ImageDuplicationDetectionService.java

```diff
@@ -0,0 +1,114 @@
+
+        log.info("Finding image duplicates for offender: '{}'", offenderNumber.getOffenderNumber());
+
+        final var imageUploads = imageUploadRepository.findByOffenderNo(offenderNumber.getOffenderNumber());
+
+        return imageUploads.stream()
+                .flatMap(this::findDuplicates)
+                .collect(toList());
+    }
+
+    private Stream<ImageDuplicate> findDuplicates(final OffenderImageUpload referenceImage) {
+
+        log.debug("Finding image duplicates for offender: '{}' and image: '{}'",
+                referenceImage.getOffenderNo(), referenceImage.getImageId());
+
+        final var matchingFaceIds = imageRecognitionClient.findMatchesFor(new FaceId(referenceImage.getFaceId()));
+
+        return matchingFaceIds.stream()
+                .map(matchingFaceId -> getImagePair(referenceImage, matchingFaceId))
+                .filter(ImagePair::haveDifferentOffenderNumbers)
+                .map(this::findOrPersistDuplicate);
+    }
+
+    private ImagePair getImagePair(final OffenderImageUpload referenceImage, final FaceId matchingFaceId) {
+
+        log.debug("Duplicate face ('{}') found for offender: '{}' and image: '{}'",
+                matchingFaceId, referenceImage.getOffenderNo(), referenceImage.getImageId());
+
+        final var matchingImage = imageUploadRepository.findByFaceId(matchingFaceId.getFaceId())
+                .orElseThrow(illegalState("Cannot find image upload for faceId: '%s'", matchingFaceId.getFaceId()));
+
+        return new ImagePair(referenceImage, matchingImage);
+    }
+
+    private ImageDuplicate findOrPersistDuplicate(final ImagePair imagePair) {
+
+        log.info("Image duplicate found for reference offender: '{}', and duplicate offender: '{}'",
+                imagePair.getReferenceOffenderNo(), imagePair.getDuplicateOffenderNo());
+
+        return imageDuplicateRepository.findByOffenderImageUploadIds(imagePair.getReferenceImageId(), imagePair.getDuplicateImageId())
+                .orElseGet(() -> persistDuplicate(imagePair.getReferenceImage(), imagePair.getDuplicateImage()));
+    }
+
+    private ImageDuplicate persistDuplicate(final OffenderImageUpload referenceImage,
+                                            final OffenderImageUpload duplicateImage) {
+        return imageDuplicateRepository.save(ImageDuplicate.builder()
+                .referenceOffenderImageUpload(referenceImage)
+                .duplicateOffenderImageUpload(duplicateImage)
+                .detectionDateTime(timeSource.nowAsLocalDateTime())
+                .build());
+    }
+
+    @Getter
+    @AllArgsConstructor
+    private static class ImagePair {
+        private final OffenderImageUpload referenceImage;
+        private final OffenderImageUpload duplicateImage;
+
+        private String getReferenceOffenderNo() {
+            return referenceImage.getOffenderNo();
+        }
+
+        private String getDuplicateOffenderNo() {
+            return referenceImage.getOffenderNo();
```

Comment:
Is this correct?  Should this be using `duplicateImage`?

Reply:
Will correct in follow-up